src/smefit/rge.py

```diff
@@ -0,0 +1,424 @@
+        if not all(scale is None for scale in dataset_scales):
+            scales.extend(dataset_scales)
+        else:
+            scales.extend([default_scale] * len(dataset_scales))
+
+        _logger.info(f"Loaded scales for dataset {dataset}: {dataset_scales}")
+
+    return scales
+
+
+def load_rge_matrix(rge_dict, coeff_list, datasets=None, theory_path=None):
+    """
+    Load the RGE matrix for the SMEFT Wilson coefficients.
+
+    Parameters
+    ----------
+    rge_dict: dict
+        dictionary with the RGE input parameter options
+    coeff_list: list
+        list of Wilson coefficients to be included in the RGE matrix
+    datasets: list
+        list of datasets
+    theory_path: str
+        path to the theory files
+
+    Returns
+    -------
+    rgemat: numpy.ndarray
+        RGE matrix
+    """
+    init_scale = rge_dict.get("init_scale", 1e3)
+    obs_scale = rge_dict.get("obs_scale", 91.1876)
+    smeft_accuracy = rge_dict.get("smeft_accuracy", "integrate")
+    adm_QCD = rge_dict.get("adm_QCD", "full")
+    yukawa = rge_dict.get("yukawa", "top")
+    rge_runner = RGE(coeff_list, init_scale, smeft_accuracy, adm_QCD, yukawa)
+    # if it is a float, it is a static scale
+    if type(obs_scale) is float or type(obs_scale) is int:
+        rgemat = rge_runner.RGEmatrix(obs_scale)
+        gen_operators = list(rgemat.index)
+        operators_to_keep = {k: {} for k in gen_operators}
+        return rgemat.values, operators_to_keep
+
+    elif obs_scale == "dynamic":
+        scales = load_scales(datasets, theory_path, default_scale=init_scale)
+
+        operators_to_keep = {}
+        rgemat = []
+        rge_cache = {}
+        for scale in scales:
+            if scale == init_scale:
+                # produce an identity matrix with row and columns coeff_list
+                rgemat_scale = pd.DataFrame(
+                    np.eye(len(coeff_list), len(coeff_list)),
+                    columns=sorted(coeff_list),
+                    index=sorted(coeff_list),
+                )
+
+            # Check if the RGE matrix has already been computed
+            elif scale in rge_cache:
+                rgemat_scale = rge_cache[scale]
+            else:
+                rgemat_scale = rge_runner.RGEmatrix(scale)
+                gen_operators = list(rgemat_scale.index)
+                # Fill with the operators if not already present in the dictionary
+                for op in gen_operators:
+                    if op not in operators_to_keep:
+                        operators_to_keep[op] = {}
+                rge_cache[scale] = rgemat_scale
+
+            rgemat.append(rgemat_scale)
+
+        # now loop through the rgemat and if there are operators
+        # that are not present in the matrix, fill them with zeros
+        for mat in rgemat:
+            for op in operators_to_keep:
+                if op not in mat.index:
+                    mat.loc[op] = np.zeros(len(mat.columns))
```

Comment:
No sorry, the UV coupling is getting a column of zeros in the RGEmatrix already, so I think it's only for the first reason, i.e. that at different scales what you get from the RGE matrix can contain a different number of operators and therefore at the end (after having the full list of operators_to_keep), you loop them all and make sure that they all contain all the operators...

Imagine the extreme case in which a RGEmatrix is computed from 1 TeV to 0.9999 TeV. This rge will basically not generate any operator and it will return as a [[1.0]]. Then you have another matrix from 1 TeV to 80 GeV. This one will generate all of them basically. So now you take the first matrix and fill it with zeros so it becomes the same size, but with a 1.0 somewhere

Reply:
You can test this with the snippet in the comment below, running to different scales, starting with the same operators, and seeing that the matrices are different.

Reply:
Ah thanks for the explanation, I'll try with the snippet you provided!